### PR TITLE
[MPS] Implement `torch.distributions.Gamma` (fwd + bwd)

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #179228

Implement `_standard_gamma` and `_standard_gamma_grad` for MPS

`_standard_gamma` uses the Marsaglia & Tsang (2000) acceptance-rejection method, ported from aten/src/ATen/native/Distributions.h which originates from NumPy's random module (Copyright 2005 Robert Kern). 
`_standard_gamma_grad` ports the three-region reparameterized gradient approximation (Taylor series, Rice saddle point, bivariate rational) from standard_gamma_grad_one in Distributions.h, reusing the existing c10::metal::digamma from special_math.h.

It was requested 10+ times on https://github.com/pytorch/pytorch/issues/141287

Implementation quality was validated using this script:

```python
"""Visual + statistical test for _standard_gamma on MPS vs CPU."""
import os
import torch
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt

alphas = [0.1, 0.5, 1.0, 2.0, 5.0, 10.0, 50.0]
N = 100_000

fig, axes = plt.subplots(2, 4, figsize=(18, 8))
axes = axes.flatten()

print(f"{'alpha':>6} | {'CPU mean':>9} | {'MPS mean':>9} | {'CPU var':>9} | {'MPS var':>9} | {'KS p-val':>8}")
print("-" * 65)

for i, a in enumerate(alphas):
    alpha = torch.full((N,), a)
    torch.manual_seed(42)
    cpu = torch._standard_gamma(alpha).numpy()
    torch.manual_seed(42)
    mps = torch._standard_gamma(alpha.to("mps")).cpu().numpy()

    _, p = stats.ks_2samp(cpu, mps)
    print(f"{a:6.1f} | {cpu.mean():9.4f} | {mps.mean():9.4f} | {cpu.var():9.4f} | {mps.var():9.4f} | {p:8.5f}")

    ax = axes[i]
    clip = np.percentile(np.concatenate([cpu, mps]), 99)
    bins = np.linspace(0, clip, 60)
    ax.hist(cpu, bins=bins, alpha=0.5, density=True, label="CPU", color="steelblue")
    ax.hist(mps, bins=bins, alpha=0.5, density=True, label="MPS", color="coral")
    x = np.linspace(1e-6, clip, 200)
    ax.plot(x, stats.gamma.pdf(x, a), "k--", lw=1.5, label="Theory")
    ax.set_title(f"α = {a}")
    ax.legend(fontsize=8)

axes[-1].axis("off")

# Gradient test
print(f"\n{'alpha':>6} | {'max |diff|':>12} | {'mean |diff|':>12}")
print("-" * 40)
for a in [0.1, 0.5, 1.0, 5.0, 50.0]:
    alpha = torch.full((10_000,), a)
    torch.manual_seed(0)
    x = torch._standard_gamma(alpha)
    d = (torch._standard_gamma_grad(alpha, x) - torch._standard_gamma_grad(alpha.to("mps"), x.to("mps")).cpu()).abs()
    print(f"{a:6.1f} | {d.max().item():12.6e} | {d.mean().item():12.6e}")

plt.suptitle("Gamma Distribution: MPS vs CPU vs Theory", fontsize=14)
plt.tight_layout()
plt.savefig("gamma_mps_vs_cpu.png", dpi=150)
```

That yields following
gamma_mps_vs_cpu

Co-authored-by: Claude <noreply@anthropic.com>

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/Distributions.metal [hunk 1]
```diff
@@ -69,3 +70,181 @@ REGISTER_OP(geometric, long);
...
+
+REGISTER_GAMMA(float);
+REGISTER_GAMMA(half);
+REGISTER_GAMMA(bfloat);
+
+// Reparameterized gradient for Gamma distribution.
+// Computes -(d/dalpha cdf(x;alpha)) / pdf(x;alpha).
+// Adapted from aten/src/ATen/native/Distributions.h standard_gamma_grad_one().
+
+constant const float GAMMA_GRAD_COEF_UV[3][8] = {
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/kernels/Distributions.metal]: Don't know if this is any different from constexpr

File: aten/src/ATen/native/mps/operations/Distributions.mm [hunk 2]
```diff
@@ -511,6 +513,75 @@ Tensor normal_mps(const Tensor& mean, const Tensor& std, std::optional<Generator
...
+Tensor _standard_gamma_grad_mps(const Tensor& self, const Tensor& output) {
+  if (self.numel() == 0) {
+    return at::empty_like(self);
+  }
+
+  using namespace mps;
+
+  auto stream = getCurrentMPSStream();
+  Tensor ret = at::empty_like(self);
+  const auto self_contig = self.is_contiguous() ? self : self.contiguous();
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/Distributions.mm]: Nit contiguous a noop if it's contiguous just call it?

**Review by Skylion007 — APPROVED** [review]

**Review by dcci — APPROVED** [review]

File: aten/src/ATen/native/mps/operations/Distributions.mm (hunk 2 above)
Comment (malfet) [reply aten/src/ATen/native/mps/operations/Distributions.mm]: LLMS :)

Comment (malfet) [reply aten/src/ATen/native/mps/operations/Distributions.mm]: But also they messed up, as ret should be `empty_like(..., contig)`
